flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java

```diff
@@ -0,0 +1,326 @@
+import org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPool;
+import org.apache.flink.runtime.jobmaster.slotpool.ResourceCounter;
+import org.apache.flink.runtime.rest.handler.legacy.utils.ArchivedExecutionGraphBuilder;
+import org.apache.flink.runtime.slots.ResourceRequirement;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+import org.slf4j.Logger;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.createSlotOffersForResourceRequirements;
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.offerSlots;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+
+/** Tests for the {@link DeclarativeScheduler}. */
+public class DeclarativeSchedulerTest extends TestLogger {
+
+    private static final int PARALLELISM = 4;
+    private static final JobVertex JOB_VERTEX;
+
+    static {
+        JOB_VERTEX = new JobVertex("v1");
+        JOB_VERTEX.setParallelism(PARALLELISM);
+        JOB_VERTEX.setInvokableClass(AbstractInvokable.class);
+    }
+
+    private final ComponentMainThreadExecutor mainThreadExecutor =
+            new ManuallyTriggeredComponentMainThreadExecutor(Thread.currentThread());
+
+    @Test
+    public void testInitialState() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(getJobGraph(), mainThreadExecutor).build();
+
+        assertThat(scheduler.getState(), instanceOf(Created.class));
+    }
+
+    @Test
+    public void testIsState() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(getJobGraph(), mainThreadExecutor).build();
+
+        final State state = scheduler.getState();
+
+        assertThat(scheduler.isState(state), is(true));
+        assertThat(scheduler.isState(new DummyState()), is(false));
+    }
+
+    @Test
+    public void testRunIfState() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(getJobGraph(), mainThreadExecutor).build();
+
+        AtomicBoolean ran = new AtomicBoolean(false);
+        scheduler.runIfState(scheduler.getState(), () -> ran.set(true));
+        assertThat(ran.get(), is(true));
+    }
+
+    @Test
+    public void testRunIfStateWithStateMismatch() throws Exception {
+        final DeclarativeScheduler scheduler =
+                new DeclarativeSchedulerBuilder(getJobGraph(), mainThreadExecutor).build();
+
+        AtomicBoolean ran = new AtomicBoolean(false);
+        scheduler.runIfState(new DummyState(), () -> ran.set(true));
+        assertThat(ran.get(), is(false));
+    }
```

Comment:
I think we are missing a case where we test that failures are forwarded to the fatal error handler.